File: Foundations of High performance computing/Assignment01/section1/summatrixfast.cc
#include <time.h>    
#include <iostream>
#include "mpi.h"
#include <vector>
#include <ctime>
#include <algorithm>
#include <random>
#include <iomanip>



int main(int argc, char **argv) {
	std::srand(unsigned(std::time(nullptr)));
	MPI_Status status;
	MPI_Request request;
	int period{1};
	int reorder{1};
	int numprocs;
	int rank;
	double sum=0;
	MPI_Init(&argc, &argv);
	double start_time, end_time;
	MPI_Comm_size(MPI_COMM_WORLD,&numprocs);
	MPI_Comm_rank(MPI_COMM_WORLD, &rank);
	int size1=std::atoi(argv[1]);
	int size2=std::atoi(argv[2]);
	int size3=std::atoi(argv[3]);
	int size=size1*size2*size3;
	int localsize=size/numprocs;
	if (size % numprocs !=  0){
	size=numprocs*ceil((double) size/numprocs);
	}	

	double global1[size];
	double global2[size];
	double global[size];
//randomizing the two matrices
for(int a=0; a<10; a++){
	start_time=MPI_Wtime();
	for (int i=0; i<size1; i++){
		for (int j=0; j<size2; j++){
			for (int k=0; k<size3; k++){
				global1[k+j*size3+i*size3*size2]=rand()%100;
			}
		}
	}
	
	for (int i=0; i<size1; i++){
                for (int j=0; j<size2; j++){
                        for (int k=0; k<size3; k++){
                        	global2[k+j*size3+i*size3*size2]=rand()%100;
                        }
                }
        }


	double local1[localsize];
	double local2[localsize];
	double local[localsize];
//Scatter
	MPI_Scatter(global1, localsize, MPI_DOUBLE, local1, localsize, MPI_DOUBLE, 0, MPI_COMM_WORLD);
	MPI_Scatter(global2, localsize, MPI_DOUBLE, local2, localsize, MPI_DOUBLE, 0, MPI_COMM_WORLD);
	
//sum of the submatrices
	for (int i=0; i<localsize; i++){
		local[i]=local1[i]+local2[i];
	}


	//gather
	MPI_Gather(local, localsize, MPI_DOUBLE, global, localsize, MPI_DOUBLE, 0, MPI_COMM_WORLD);
	
	end_time=MPI_Wtime();
	sum=sum+end_time-start_time;
}
	sum=sum/10;
	if(rank==0){
		std::cout<<std::fixed<<std::setprecision(2)<<end_time-start_time<<std::endl;
        }

	
	MPI_Finalize();
	return 0;
}

File: Foundations of High performance computing/Assignment01/section1/summatrixfast.py
from mpi4py import MPI
import numpy as np
import sys
import math
comm = MPI.COMM_WORLD
status = MPI.Status()
request= MPI.Request()
numprocs = comm.Get_size()
rank= comm.Get_rank()
size1=int(sys.argv[1])
size2=int(sys.argv[2])
size3=int(sys.argv[3])
size=size1*size2*size3
flag=0
sum=0
localsize=int(size/numprocs);
if size%numprocs!=0:
	localsize=int(math.ceil(size/numprocs))
	flag=1	
	newsize=localsize*numprocs
	
for a in range(5):
	start_time = MPI.Wtime()
	global1=np.random.rand(size1,size2,size3)
	global2=np.random.rand(size1,size2,size3)
	if rank==0:
		if flag==1:
#Padding if necessary
			global1.shape=(size,)
			global2.shape=(size,)
			np.pad(global1, (0, newsize-size), 'constant')
			np.pad(global2, (0, newsize-size), 'constant')

	local1=np.zeros(localsize)	
	local2=np.zeros(localsize)

	if flag==0:
		global3=np.zeros([size1,size2,size3])
	else:
		global3=np.zeros(newsize);
#scatter 
	comm.Scatter([global1, localsize, MPI.DOUBLE], local1, root=0)
	comm.Scatter([global2, localsize, MPI.DOUBLE], local2, root=0)
#sum
	local=np.add(local1,local2)
#
	comm.Gather([local, localsize, MPI.DOUBLE], global3, root=0)
#restore matrix to original size
	if flag==1:
		global3=global3[0:(size)]
		global1=global1[0:(size)]
		global2=global2[0:(size)]
		global3.shape=(size1,size2,size3)
		global2.shape=(size1,size2,size3)
		global1.shape=(size1,size2,size3)	
	end_time = MPI.Wtime()
	sum=sum+end_time-start_time

sum=sum/5	
if rank==0:
	print("{:.2f}".format(sum))



